### PR TITLE
fix: use default total send fee provided by the gateway

### PR DESCRIPTION
**Pull request by joschisan**
The reasoning here is that the gateway is supposed to choose a fee for sending an outgoing payment that is sufficient for high lightning reliability; retrying with higher fees in case of cancellation is unnecessary. 

Furthermore, the fee behaviour is now identical to receiving an incoming payment.

### CODE REVIEW

**Comment by elsirion** [comment]
Why unnecessarily limit options? This will just be a headache compatibility-wise once we want to introduce a more dynamic protocol for whatever reason.

**Comment by joschisan** [comment]
> Why unnecessarily limit options? This will just be a headache compatibility-wise once we want to introduce a more dynamic protocol for whatever reason.

If you look into the pr, you will find that this is only a client side change.

The Idea here for the send fee (and possibly also the time lock expiration delta) is the following. 

Successfully routing a payment depends on many factors, two of which are maximum fee and maximum cults expiry. If the gateway only sets a minimum fee for itself and a minimum expiration delta required for the gateway to claim the payment, it is left to the user to guess how much additional fee and expiration delta is required to route a payment successfully. Furthermore, appropriate settings for both depend on the gateways connectivity and may vary between gateways. Hence the user either chooses a high default for both (which we do right now) or the user may has to do multiple cancelled attempts with different fees and expiration deltas.

Alternatively, if we assume that reliability, meaning payments are successfully routed on the first try, is our priority, we can let the gateway set higher minimums for send fee and expiration delta, that includes a fee and time lock delta for lightning such that, if the payment still cannot be routed, it can be assumed by the user that the failure was very likely not caused by insufficient fees or expiration delta. Retrying with higher fees and expiration delta is very likely to fail again, the user should try other gateways if available or consider the invoice unpayable, receiver may be offline for example.

So the statement the gateway makes with the returned fee and expiration delta is 

"this is the minimum fee and expiration delta I require, please balance reliability, fees and timeout on your own" 

but instead 

"I can provide a reliable service for this fee and expiration delta"

which seems to me to be more appropriate to the less technical end user we generally assume.

Thoughts?

**Comment by joschisan** [comment]
@elsirion and I decided to client-side rely on a default fee returned by the gateway, however we still leave the option for a custom fee such that a fee-sensitive user may probe for a specific invoice before setting a send fee.

File: gateway/ln-gateway/src/gateway_module_v2/send_sm.rs [hunk 1]
```diff
@@ -137,18 +137,11 @@ impl SendStateMachine {
...
-            .add_fee(invoice_msats);
-
-        if contract_amount < min_contract_amount {
+        if contract.amount < min_contract_amount {
             return Err(Cancelled::Underfunded);
         }
 
-        let excess_fee = contract_amount - min_contract_amount;
-        let max_fee_msat = excess_fee.msats + (min_contract_amount.msats - invoice_msats) / 2;
+        let max_fee_msat = (contract.amount - min_contract_amount).msats;
```
Comment (elsirion) [inline gateway/ln-gateway/src/gateway_module_v2/send_sm.rs:144]: nit: `checked_sub` could do both in one

File: gateway/ln-gateway/src/lib.rs [hunk 2]
```diff
@@ -1631,6 +1622,13 @@ impl Gateway {
             .amount_milli_satoshis()
             .ok_or(anyhow!("Invoice is missing amount"))?;
 
+        let min_contract_amount = self
+            .payment_info_v2(&payload.federation_id)
+            .await
+            .ok_or(anyhow!("Payment Info not available"))?
+            .send_fee_minimum
+            .add_fee(invoice_msats);
```
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:1630]: Not sure why you moved that here, but you probably had a good reason

File: modules/fedimint-lnv2-client/src/lib.rs [hunk 3]
```diff
@@ -134,16 +134,12 @@ pub struct SendPaymentPayload {
 #[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Decodable, Encodable)]
 pub struct PaymentInfo {
     pub public_key: PublicKey,
-    pub payment_fees: PaymentFees,
+    pub send_fee_minimum: PaymentFee,
+    pub send_fee_default: PaymentFee,
+    pub receive_fee: PaymentFee,
```
Comment (elsirion) [inline modules/fedimint-lnv2-client/src/lib.rs:139]: :+1: could use some docs on what these two send fees mean I guess

**Review by elsirion — APPROVED** [review]

**Pull request merged**